tcpdump.c

```diff
@@ -0,0 +1,239 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include <string.h>
+#include <errno.h>
+#include <net/if.h>
+#include <linux/if_ether.h>
+#include <linux/if_packet.h>
+#include <linux/filter.h>
+#include <arpa/inet.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <net/if_arp.h>
+#include <pcap/dlt.h>
+
+#include "error.h"
+#include "bpf_pcap.h"
+#include "pcap_file.h"
+
+// TODO: things i'm not sure about:
+// 1. after create_raw_socket should i pass the sock_fd or a pointer to the sock_fd (right now its sock_fd)
+// 2. if a CHECK fails should i close the socket in that function or in 
+// an upper function(the function that create the socket)
+// 3. how do i create a good CHECK with custom error message? should i just pass the error message?
+
+// linux MTU (including loopback interface) defaults to 65536
+// to handle MSG_TRUNC without using recvmsg (and creating msghdr struct) we define the
+// size to be 1 byte bigger and check if a recieved packet is equal to MAX_PKT_LEN
+// snapshot len should then be MAX_PKT_LEN -1 
+#define MAX_PKT_LEN (65537)
+// TODO: fix, this is kinda random
+#define MAX_BPF_LEN (50)
+
+static bool running = true;
+
+void sigint_handler(int sig) {
+
+	running = false;
+    return;
+
+}
+
+static error_status_t set_bpf(int sock, int link_type, const char* bpf_str, int opt) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    struct sock_fprog bpf_filter = {0};
+    struct sock_filter* bpf_buffer = NULL;
+    
+    CHECK(sock != -1);
+    CHECK(bpf_str != NULL);
+
+    bpf_buffer = malloc(sizeof(struct sock_filter) * MAX_BPF_LEN);
+    CHECK(bpf_buffer != NULL);
+
+    bpf_filter.filter = bpf_buffer;
+
+    CHECK_FUNC(compile_bpf(MAX_PKT_LEN - 1, link_type, &bpf_filter, bpf_str, opt));
+
+    // for debug purpose
+    // the 2 stands for how i print the bpf 
+    dump_bpf(&bpf_filter, 2); 
+    CHECK(setsockopt(sock, SOL_SOCKET, SO_ATTACH_FILTER, &bpf_filter, sizeof(bpf_filter)) >= 0);
+
+cleanup:
+    free(bpf_buffer); // Best effort.
+    return ret_status;
+}
+
+static error_status_t iface_name_to_index(int sock, const char* if_name, int* if_index) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    struct ifreq ifr = {0};
+    size_t if_name_len = 0;
+
+    CHECK(if_name != NULL);
+    CHECK(if_index != NULL);
+    
+    if_name_len = strlen(if_name);
+
+    CHECK_STR(if_name_len < sizeof(ifr.ifr_name), "Interface name too long");
```

Comment:
Why not <=
Also, this is better than <=IFNAMSIZ in bpf_pcap.c:28 - change it there :)